tensorflow_probability/python/optimizer/bfgs.py

```diff
@@ -604,12 +604,12 @@ def _do_update_fn():
     """Updates the Hessian estimate."""
     # The quadratic form: y^T.H.y.
     n = len(grad_delta.shape.as_list())
```

Comment:
We would prefer to have full support for dynamic shapes which includes statically unknown rank as well. shape.as_list() will raise an exception if shape = None.

Reply:
Hmm. Do you know what the best way of getting the number of dimensions of a tensor with unknown shape is? I think `tf.shape(tf.shape(grad_delta))` will work, although it's kind of ugly. Passing in a tensor there might be enough if the input shape is already `<unknown>`, since the only thing that matters is preserving the shape info through the tensordot.

Reply:
[tf.rank()](https://www.tensorflow.org/api_docs/python/tf/rank) will give you the dynamic rank of the tensor.